src/evented.rs

```diff
@@ -0,0 +1,146 @@
+use event::Event;
+use std::collections::VecDeque;
+use std::fmt::Debug;
+use std::mem;
+
+#[macro_export]
+macro_rules! try_ev {
+    ($val:expr, $events:expr) => {
+        match $val {
+            Ok(v) => v,
+            Err(err) => return $events.with_value(Err(From::from(err)))
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! try_evx {
+    ($val:expr, $events:expr) => {
+        try_ev!($val.extract(&mut $events), $events)
+    }
+}
+
+/// An Evented<T> wraps a T and bundles some events alongside it.
+///
+/// The public methods on Evented<T> are such that it is difficult
+/// to drop the events contained.
+#[must_use]
+pub struct Evented<T, Ev = Event>
+    where Ev: Debug
+{
+    events: Vec<Ev>,
+    /// The value will always be Some(t: T) unless this struct is about to be dropped.
+    value: Option<T>,
+}
+
+/// Trait for trivial conversions to Evented
+pub trait ToEvented<Ev>: Sized
+    where Ev: Debug
+{
+    /// Wrap self in an evented type with no events.
+    fn to_evented(self) -> Evented<Self, Ev> {
+        Evented::new(vec![], self)
+    }
+}
+
+/// Blanket impl of `ToEvented` for every sized type.
+impl<T, Ev: Debug> ToEvented<Ev> for T where T: Sized {}
+
+/// Explode if an attempt is made to drop events.
+impl<T, Ev: Debug> Drop for Evented<T, Ev> {
+    fn drop(&mut self) {
+        if !self.events.is_empty() {
+            error!("Events were dropped: {:?}", self.events);
+            panic!();
```

Comment:
The idea behind this panic is that dropping events should be considered a fatal bug... however, panicking in a `Drop` implementation will likely abort (as per the [docs](https://doc.rust-lang.org/std/ops/trait.Drop.html#panics)), and indeed in practice I found it would generate a `SIGILL` on my Mac.

An alternative would be to just print the error and not panic at all, but I worry this would make it too easy to miss an event drop bug?